src/test/java/AuthenticatedHttpServletTest.java

```diff
@@ -0,0 +1,201 @@
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mockito;
+
+@RunWith(JUnit4.class)
+public class AuthenticatedHttpServletTest {
+  // Mock abstract servlet with real constructor and real methods
+  private static final AuthenticationVerifier authVerifier =
+      Mockito.mock(AuthenticationVerifier.class);
+  private static final AuthenticatedHttpServlet servlet =
+      Mockito.mock(
+          AuthenticatedHttpServlet.class,
+          Mockito.withSettings()
+              .useConstructor(authVerifier)
+              .defaultAnswer(Mockito.CALLS_REAL_METHODS));
+
+  private HttpServletRequest request;
+  private HttpServletResponse response;
+  private StringWriter stringWriter;
+  private PrintWriter printWriter;
+
+  private static final Boolean AUTHENTICATION_VERIFIED = true;
+  private static final Boolean AUTHENTICATION_NOT_VERIFIED = false;
+
+  private static final String ID_TOKEN_KEY = "idToken";
+  private static final String ID_TOKEN_VALUE = "sampleId";
+  private static final String ACCESS_TOKEN_KEY = "accessToken";
+  private static final String ACCESS_TOKEN_VALUE = "sampleAccessToken";
+
+  private static final Cookie sampleIdTokenCookie = new Cookie(ID_TOKEN_KEY, ID_TOKEN_VALUE);
+  private static final Cookie sampleAccessTokenCookie =
+      new Cookie(ACCESS_TOKEN_KEY, ACCESS_TOKEN_VALUE);
+  private static final Cookie emptyIdToken = new Cookie(ID_TOKEN_KEY, "");
+  private static final Cookie emptyAccessToken = new Cookie(ACCESS_TOKEN_KEY, "");
+
+  private static final Cookie[] noCookies = new Cookie[] {};
+  private static final Cookie[] noAccessTokenCookies = new Cookie[] {sampleIdTokenCookie};
+  private static final Cookie[] noIdTokenCookies = new Cookie[] {sampleAccessTokenCookie};
+  private static final Cookie[] emptyIdTokenCookies =
+      new Cookie[] {emptyIdToken, sampleAccessTokenCookie};
+  private static final Cookie[] emptyAccessTokenCookies =
+      new Cookie[] {sampleIdTokenCookie, emptyAccessToken};
+  private static final Cookie[] validCookies =
+      new Cookie[] {sampleIdTokenCookie, sampleAccessTokenCookie};
+
+  @Before
+  public void init() throws IOException {
+    request = Mockito.mock(HttpServletRequest.class);
+    response = Mockito.mock(HttpServletResponse.class);
+
+    // Writer used in get/post requests to capture HTTP response values
+    stringWriter = new StringWriter();
+    printWriter = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(printWriter);
+  }
+
+  @After
+  public void clear() {
+    // Dump contents after each test
+    stringWriter.getBuffer().setLength(0);
+  }
+
+  @Test
+  public void getRequestNoTokens() throws IOException {
+    when(request.getCookies()).thenReturn(noCookies);
+    servlet.doGet(request, response);
+    verify(response, Mockito.times(1)).sendError(Mockito.eq(403), Mockito.anyString());
```

Comment:
Just out of curiosity, what does `Mockito.times(1)` do in the context of this line?

Reply:
It verifies that the method was called exactly once :)